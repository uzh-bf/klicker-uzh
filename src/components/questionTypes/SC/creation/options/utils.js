// reorder an array using start and end index
const reorder = (list, startIndex, endIndex) => {
  const result = Array.from(list)
  const [removed] = result.splice(startIndex, 1)
  result.splice(endIndex, 0, removed)

  return result
}

// handle the drag end event
const onDragEnd = (items, result, cb) => {
  // check whether the item was dropped outside the list
  if (!result.destination) {
    return
  }

  // reorder the items accordingly
  const reordered = reorder(items, result.source.index, result.destination.index)

  // callback with the new order
  cb(reordered)
}

export { reorder, onDragEnd }
