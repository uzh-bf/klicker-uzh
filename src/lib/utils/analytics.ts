import ReactGA from 'react-ga'

export function initGA(id: string): void {
  ReactGA.initialize(id)
}

export function logPageView(): void {
  ReactGA.set({ page: window.location.pathname })
  ReactGA.pageview(window.location.pathname)
}

export function logEvent(category: string = '', action: string = ''): void {
  if (category && action) {
    ReactGA.event({ action, category })
  }
}

export function logException(description: string = '', fatal: boolean = false): void {
  if (description) {
    ReactGA.exception({ description, fatal })
  }
}
