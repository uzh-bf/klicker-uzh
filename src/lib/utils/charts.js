// import _round from 'lodash/round'

import {
  // QUESTION_TYPES,
  SMALL_BAR_THRESHOLD,
  SMALL_PIE_THRESHOLD,
  CHART_TYPES,
} from '../../constants'

// convert an integer index into letters
export function indexToLetter(index) {
  // calculate how many times the letter should be repeated
  // eg. 0 => A (times= 1), 24 => AA (times = 2),
  const n = Math.floor(index / 26) + 1

  // index: letter
  // 65: A, 66: B , ...
  // return n times a letter from A to Z
  return Array(n + 1).join(String.fromCharCode(65 + (index % 26)))
}

// determine whether the label (A,B, ...) is displayed within the bar or not
export function getLabelIn(
  chartType,
  questionType,
  count,
  totalResponses,
  index,
) {
  if (chartType === CHART_TYPES.BAR_CHART) {
    if (count / totalResponses > SMALL_BAR_THRESHOLD) {
      return count
    }
  }

  if (chartType === CHART_TYPES.PIE_CHART) {
    if (count / totalResponses > SMALL_PIE_THRESHOLD) {
      return `${indexToLetter(index)}`
    }
  }

  return ''
}

// determine whether the label (A,B, ...) is displayed outside the bar or not
export function getLabelOut(
  chartType,
  questionType,
  count,
  totalResponses,
  index,
) {
  if (chartType === CHART_TYPES.BAR_CHART) {
    if (count / totalResponses <= SMALL_BAR_THRESHOLD) {
      return count
    }

    return ''
  }

  if (chartType === CHART_TYPES.PIE_CHART) {
    if (count / totalResponses <= SMALL_PIE_THRESHOLD) {
      return `${count} (${indexToLetter(index)})`
    }

    return count
  }

  return ''
}
