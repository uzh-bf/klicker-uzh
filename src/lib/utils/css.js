// @flow

/* eslint-disable import/prefer-default-export */

import React from 'react'

function createLinks(links: Array<string> = []): Array<*> {
  return links.map(link => (
    <link
      key={link}
      rel="stylesheet"
      href={
        link.substr(0, 4) === 'http' ? (
          link
        ) : (
          `https://cdnjs.cloudflare.com/ajax/libs/semantic-ui/2.2.11/components/${link}.min.css`
        )
      }
    />
  ))
}

export { createLinks }
