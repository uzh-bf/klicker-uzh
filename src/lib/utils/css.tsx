/* eslint-disable import/prefer-default-export */

import React from 'react'

export function createLinks(links: string[] = []): React.ReactElement[] {
  return links.map(link => <link href={link} key={link} rel="stylesheet" />)
}

