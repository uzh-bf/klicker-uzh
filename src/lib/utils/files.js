import axios from 'axios'

export function getPresignedURLs(files, requestPresignedURL) {
  return Promise.all(
    files.map(async file => {
      const result = await requestPresignedURL({
        variables: {
          fileType: file.type,
        },
      })
      const { fileName, signedUrl } = result.data.requestPresignedURL
      return { file, fileName, signedUrl }
    })
  )
}

export function uploadFilesToPresignedURLs(files) {
  return Promise.all(files.map(({ file, signedUrl }) => axios.put(signedUrl, file)))
}
