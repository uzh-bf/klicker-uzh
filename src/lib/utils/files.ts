import axios from 'axios'

export function getPresignedURLs(files: any[], requestPresignedURL: any): Promise<any> {
  return Promise.all(
    files.map(async file => {
      const result = await requestPresignedURL({
        variables: {
          fileType: file.type,
        },
      })
      const { fileName, signedUrl } = result.data.requestPresignedURL
      return { file, fileName, signedUrl }
    })
  )
}

export function uploadFilesToPresignedURLs(files: any[]): Promise<any> {
  return Promise.all(files.map(({ file, signedUrl }) => axios.put(signedUrl, file)))
}
