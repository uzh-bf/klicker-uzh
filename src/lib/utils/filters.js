// @flow
/* eslint-disable no-unused-vars */

import _every from 'lodash/every'

import type { QuestionType } from '../../queries/queries'

type QuestionFilters = {
  tags: Array<string>,
  title: ?string,
  type: ?string,
}

function filterQuestions(questions: QuestionType[], filters: QuestionFilters) {
  return questions.filter((question) => {
    if (filters.title && !question.title.includes(filters.title)) {
      return false
    }
    if (
      filters.tags &&
      !_every(filters.tags, tag => question.tags.map(t => t.name).includes(tag))
    ) {
      return false
    }
    if (filters.type && question.type !== filters.type) {
      return false
    }
    return true
  })
}

function filterSessions(sessions: any, filters: any) {
  return sessions
}

export { filterQuestions, filterSessions }
export type { QuestionFilters }
