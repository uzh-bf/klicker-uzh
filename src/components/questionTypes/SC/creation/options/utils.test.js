import { reorder, onDragEnd } from './utils'

expect.addSnapshotSerializer({
  print: val => val.name,
  test: val => val.name,
})

const options = [{ name: 'option 1' }, { name: 'option 2' }, { name: 'option 3' }]

describe('reorder', () => {
  it('updates order correctly', () => {
    // [1, 2, 3] => [2, 3, 1]
    const options2 = reorder(options, 0, 2)
    expect(options2).toMatchSnapshot()

    // [2, 3, 1] => [1, 2, 3]
    const options3 = reorder(options2, 2, 0)
    expect(options3).toMatchSnapshot()

    // [1, 2, 3] => [3, 1, 2]
    const options4 = reorder(options3, 2, 0)
    expect(options4).toMatchSnapshot()

    // [3, 1, 2] => [2, 3, 1]
    const options5 = reorder(options4, 2, 0)
    expect(options5).toMatchSnapshot()
  })

  it('handles overflowing indices', () => {
    // TODO
  })
})

describe('onDragEnd', () => {
  // [1, 2, 3] => [2, 3, 1]
  const result = { destination: { index: 2 }, source: { index: 0 } }

  it('handles dragging an item correctly', () => {
    onDragEnd(options, result, (reordered) => {
      expect(reordered).toMatchSnapshot()
    })
  })

  it('handles items dropped outside the dropzone', () => {
    onDragEnd(options, { ...result, destination: null }, (reordered) => {
      expect(reordered).toBeNull()
    })
  })
})
