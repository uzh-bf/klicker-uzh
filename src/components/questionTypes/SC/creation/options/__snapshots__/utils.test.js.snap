// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onDragEnd handles dragging an item correctly 1`] = `
Array [
  option 2,
  option 3,
  option 1,
]
`;

exports[`reorder updates order correctly 1`] = `
Array [
  option 2,
  option 3,
  option 1,
]
`;

exports[`reorder updates order correctly 2`] = `
Array [
  option 1,
  option 2,
  option 3,
]
`;

exports[`reorder updates order correctly 3`] = `
Array [
  option 3,
  option 1,
  option 2,
]
`;

exports[`reorder updates order correctly 4`] = `
Array [
  option 2,
  option 3,
  option 1,
]
`;
